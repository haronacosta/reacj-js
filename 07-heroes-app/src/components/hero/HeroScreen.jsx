import { useMemo } from 'react';
import { useParams, Navigate, useNavigate } from 'react-router-dom';

import { getHeroById } from '../../selectors/getHeroById';

const HeroScreen = () => {
  const { id } = useParams();

  const navigate = useNavigate();

  const hero = useMemo(() => getHeroById(id), [id]);

  if (!hero) {
    return <Navigate to='/' />;
  }

  const handleReturn = () => {
    //Back to the previous page
    navigate(-1);
  };

  const imagePath = `/assets/heroes/${id}.jpg`;

  const { superhero, alter_ego, characters, first_appearance, publisher } =
    hero;

  return (
    <div className='row mt-5'>
      <div className='col-4'>
        <img
          src={imagePath}
          alt={hero.superhero}
          className='img-thumbnail animate__animated animate__fadeInLeft'
        />
      </div>

      <div className='col-8'>
        <h3>{superhero}</h3>

        <ul className='list-group'>
          <li className='list-group-item'>
            <b>Alter ego: </b>
            {alter_ego}
          </li>

          <li className='list-group-item'>
            <b>Publisher: </b>
            {publisher}
          </li>

          <li className='list-group-item'>
            <b>First Appearance: </b>
            {first_appearance}
          </li>
        </ul>

        {characters && (
          <>
            <h5 className='mt-4'>Characters: </h5>
            <p>{characters}</p>
          </>
        )}

        <button className='btn btn-outline-info' onClick={handleReturn}>
          Back
        </button>
      </div>
    </div>
  );
};

export default HeroScreen;
