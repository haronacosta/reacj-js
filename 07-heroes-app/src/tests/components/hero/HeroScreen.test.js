import { mount } from 'enzyme';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import HeroScreen from '../../../components/hero/HeroScreen';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('Pruebas en el componete <HeroScreen/>', () => {
  test('No debe mostrar el <HeroScreen /> si no hay un heroe', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero']}>
        <Routes>
          <Route path='/hero' element={<HeroScreen />} />
          <Route path='/' element={<h1>No hero page</h1>} />
        </Routes>
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('h1').text().trim()).toBe('No hero page');
  });

  test('Debe de mostrar el heroe si el existe', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-spider']}>
        <Routes>
          <Route path='/hero/:id' element={<HeroScreen />} />
          <Route path='/' element={<h1>No hero page</h1>} />
        </Routes>
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.heroName').text().trim()).toBe('Spider Man');
  });

  test('debe de regresar a la pantalla anterior', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-spider']}>
        <Routes>
          <Route path='/hero/:id' element={<HeroScreen />} />
          <Route path='/' element={<h1>No hero page</h1>} />
        </Routes>
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();

    const handleClick = wrapper.find('.btn-outline-info').prop('onClick');

    handleClick();

    expect(mockNavigate).toHaveBeenCalledWith(-1);
  });

  test('Debe mostrar el no page si el heroe no existe', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-spider121212']}>
        <Routes>
          <Route path='/hero/:id' element={<HeroScreen />} />
          <Route path='/' element={<h1>No hero page</h1>} />
        </Routes>
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find('h1').text().trim()).toBe('No hero page');
  });
});
