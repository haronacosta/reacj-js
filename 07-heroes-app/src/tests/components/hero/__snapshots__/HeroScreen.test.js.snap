// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componete <HeroScreen/> Debe de mostrar el heroe si el existe 1`] = `
<div
  className="row mt-5"
>
  <div
    className="col-4"
  >
    <img
      alt="Spider Man"
      className="img-thumbnail animate__animated animate__fadeInLeft"
      src="/assets/heroes/marvel-spider.jpg"
    />
  </div>
  <div
    className="col-8"
  >
    <h3
      className="heroName"
    >
      Spider Man
    </h3>
    <ul
      className="list-group"
    >
      <li
        className="list-group-item"
      >
        <b>
          Alter ego: 
        </b>
        Peter Parker
      </li>
      <li
        className="list-group-item"
      >
        <b>
          Publisher: 
        </b>
        Marvel Comics
      </li>
      <li
        className="list-group-item"
      >
        <b>
          First Appearance: 
        </b>
        Amazing Fantasy #15
      </li>
    </ul>
    <h5
      className="mt-4"
    >
      Characters: 
    </h5>
    <p>
      Peter Parker
    </p>
    <button
      className="btn btn-outline-info"
      onClick={[Function]}
    >
      Back
    </button>
  </div>
</div>
`;

exports[`Pruebas en el componete <HeroScreen/> Debe mostrar el no page si el heroe no existe 1`] = `
<h1>
  No hero page
</h1>
`;

exports[`Pruebas en el componete <HeroScreen/> No debe mostrar el <HeroScreen /> si no hay un heroe 1`] = `
<h1>
  No hero page
</h1>
`;

exports[`Pruebas en el componete <HeroScreen/> debe de regresar a la pantalla anterior 1`] = `
<div
  className="row mt-5"
>
  <div
    className="col-4"
  >
    <img
      alt="Spider Man"
      className="img-thumbnail animate__animated animate__fadeInLeft"
      src="/assets/heroes/marvel-spider.jpg"
    />
  </div>
  <div
    className="col-8"
  >
    <h3
      className="heroName"
    >
      Spider Man
    </h3>
    <ul
      className="list-group"
    >
      <li
        className="list-group-item"
      >
        <b>
          Alter ego: 
        </b>
        Peter Parker
      </li>
      <li
        className="list-group-item"
      >
        <b>
          Publisher: 
        </b>
        Marvel Comics
      </li>
      <li
        className="list-group-item"
      >
        <b>
          First Appearance: 
        </b>
        Amazing Fantasy #15
      </li>
    </ul>
    <h5
      className="mt-4"
    >
      Characters: 
    </h5>
    <p>
      Peter Parker
    </p>
    <button
      className="btn btn-outline-info"
      onClick={[Function]}
    >
      Back
    </button>
  </div>
</div>
`;
